modules/generic/generic.go

```diff
@@ -0,0 +1,165 @@
+	return controller.Host().Config().Scope("modules").Get(controller.Name()).PopulateStruct(config)
+}
+
+// Controller holds data for a Module, and provides functionality
+// for the Module to notify the wrapping service.Module of state changes.
+// TODO(pedge): not the best name
+// TODO(pedge): split up into read-only and read-write interfaces?
+type Controller interface {
+	Host() service.Host
+	Roles() []string
+	Name() string
+	Tracer() opentracing.Tracer
+	Log() ulog.Log
+	// NotifyStopped will result in the wrapping service.Module's
+	// IsRunning() method to return false.
+	NotifyStopped()
+}
+
+// Module is a simpliciation of the service.Module interface
+// that can be wrapped with a service.Module for easier implemenation.
+type Module interface {
+	// This will be called after the Controller is populated.
+	// If the module wishes, this can be stored for use in the module.
+	//
+	// This should only be called by the generic package, and unless there are
+	// other calls, it can be safely assumed that this will be called exactly once,
+	// and called before Start() and Stop() are ever called.
+	Initialize(controller Controller) error
+	// Start the module.
+	// On return,  the module is expected to be started, unless there is an error.
+	Start() error
+	// Stop the module.
+	// On return the module is expected to be stopped, regardless of if there is an error.
+	Stop() error
+}
+
+// NewModule returns a ModuleCreateFunc for the given Module.
+func NewModule(
+	moduleName string,
+	module Module,
+	options ...modules.Option,
+) service.ModuleCreateFunc {
+	return func(moduleCreateInfo service.ModuleCreateInfo) ([]service.Module, error) {
+		module, err := newWrapperModule(moduleCreateInfo, moduleName, module, options...)
+		if err != nil {
+			return nil, err
+		}
+		return []service.Module{module}, nil
+	}
+}
+
+type wrapperModule struct {
+	modules.ModuleBase
+	log        ulog.Log
+	moduleName string
+	module     Module
+	lock       sync.RWMutex
+	running    bool
+}
+
+func newWrapperModule(
+	moduleCreateInfo service.ModuleCreateInfo,
+	moduleName string,
+	module Module,
+	options ...modules.Option,
+) (service.Module, error) {
+	for _, option := range options {
+		if err := option(&moduleCreateInfo); err != nil {
+			return nil, err
+		}
+	}
+	if moduleCreateInfo.Name != "" {
+		moduleName = moduleCreateInfo.Name
+	}
+	moduleBase := *modules.NewModuleBase(
+		moduleName,
+		moduleCreateInfo.Host,
+		moduleCreateInfo.Roles,
+	)
+	wrapperModule := &wrapperModule{
+		ModuleBase: moduleBase,
+		log:        ulog.Logger().With("moduleName", moduleName),
+		moduleName: moduleName,
+		module:     module,
+	}
+	if err := module.Initialize(wrapperModule); err != nil {
+		return nil, err
+	}
+	return wrapperModule, nil
+}
+
+func (m *wrapperModule) Name() string {
+	return m.moduleName
+}
+
+func (m *wrapperModule) Start(readyC chan<- struct{}) <-chan error {
+	m.lock.Lock()
+	defer m.lock.Unlock()
```

Comment:
I don't think it has much cost, and it is proper. I don't think of locking vs not as "overkill", its about what is correct - without it, wrapperModule isn't thread-safe